src/Psy/Readline/TabCompletion.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Psy\Readline;
+
+use Psy\Context;
+
+/**
+ * Class TabCompletion
+ * @package Psy\Readline
+ */
+class TabCompletion
+{
+    /** @var Context */
+    protected $context;
+
+    public function __construct(Context $context = null)
+    {
+        $this->context = $context;
+    }
+
+    public function activate()
+    {
+        readline_completion_function(array(&$this, 'callback'));
+    }
+
+    /**
+     * @param $input
+     * @param $index
+     * @return array
+     */
+    public function callback($input, $index)
+    {
+        $info = readline_info();
+
+        // the char just before the cursor is a dollar? send a variable context
+        $charAt = substr($info['line_buffer'], $index-1, 1);
+        if ($charAt == '$') {
+            return array_keys($this->context->getAll());
+        }
+
+        $parenthesize = function ($name) {
+            return sprintf('%s()', $name);
+        };
+
+        if (strlen($info['line_buffer']) > 4) {
+            // if the current position of the cursor has a precending new keyword send the classes names
+            $newKeyword = substr($info['line_buffer'], $index - 4, 3);
+            if ($newKeyword == 'new') {
+                return array_map($parenthesize, array_filter(get_declared_classes(), function ($class) use ($input) {
+                    return preg_match(sprintf('#^%s#', $input), $class);
+                }));
+            }
+        }
+
+        // for everything else send the functions names
+        $functions = get_defined_functions();
+        return array_map(
+            $parenthesize,
+            array_merge(
+                $functions['internal'],
+                $functions['user']
```

Comment:
Nice! Of those, it seems like these would make the most sense:

```
'array'
'clone'
'declare'
'die'
'echo'
'empty'
'eval'
'exit'
'include'
'include_once'
'isset'
'list'
'print'
'require'
'require_once'
'unset'
```

Those are the most function-like of them, and the ones that I would be most likely to expect to have completed. There are a few more that are function-like but don't make much sense in a REPL context, and the rest of them are either complex control structures, or things that would need to be completed without parens, so let's hold off on them.